moquette-0.16/broker/src/main/java/io/moquette/broker/security/ResourceAuthenticator.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * To encode your password from command line on Linux systems, you could use:
+ *
+ * <pre>
+ *     echo -n "yourpassword" | sha256sum
+ * </pre>
+ *
+ * NB -n is important because echo append a newline by default at the of string. -n avoid this
+ * behaviour.
+ */
+public class ResourceAuthenticator implements IAuthenticator {
+
+    protected static final Logger LOG = LoggerFactory.getLogger(ResourceAuthenticator.class);
+
+    private Map<String, String> m_identities = new HashMap<>();
+
+    public ResourceAuthenticator(IResourceLoader resourceLoader, String resourceName) {
+        try {
+            MessageDigest.getInstance("SHA-256");
+        } catch (NoSuchAlgorithmException nsaex) {
+            LOG.error("Can't find SHA-256 for password encoding", nsaex);
+            throw new RuntimeException(nsaex);
+        }
+
+        LOG.info(String.format("Loading password %s %s", resourceLoader.getName(), resourceName));
+        Reader reader = null;
+        try {
+            reader = resourceLoader.loadResource(resourceName);
+            if (reader == null) {
+                LOG.warn(String.format("Parsing not existing %s %s", resourceLoader.getName(), resourceName));
+            } else {
+                parse(reader);
+            }
+        } catch (IResourceLoader.ResourceIsDirectoryException e) {
+            LOG.warn(String.format("Trying to parse directory %s", resourceName));
+        } catch (ParseException pex) {
+            LOG.warn(
+                    String.format("Format error in parsing password %s %s", resourceLoader.getName(), resourceName),
+                    pex);
+        }
+    }
+
+    private void parse(Reader reader) throws ParseException {
+        if (reader == null) {
+            return;
+        }
+
+        BufferedReader br = new BufferedReader(reader);
+        String line;
+        try {
+            while ((line = br.readLine()) != null) {
+                int commentMarker = line.indexOf('#');
+                if (commentMarker != -1) {
+                    if (commentMarker == 0) {
+                        // skip its a comment
+                        continue;
+                    } else {
+                        // it's a malformed comment
+                        throw new ParseException(line, commentMarker);
+                    }
+                } else {
+                    if (line.isEmpty() || line.matches("^\\s*$")) {
+                        // skip it's a black line
+                        continue;
+                    }
+
+                    // split till the first space
+                    int delimiterIdx = line.indexOf(':');
+                    String username = line.substring(0, delimiterIdx).trim();
+                    String password = line.substring(delimiterIdx + 1).trim();
+
+                    m_identities.put(username, password);
+                }
+            }
+        } catch (IOException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)